File: packages/web/src/utils/__tests__/ownershipInsights.test.ts
import { describe, expect, it } from 'vitest';
import { computeOwnershipInsights } from '../ownershipInsights';

describe('ownershipInsights', () => {
  it('derives finance payment, principal, depreciation, and payment schedule', () => {
    const currentYear = new Date().getFullYear();
    const categories = [
      {
        key: 'finance',
        items: [
          {
            files: [
              {
                analysis: {
                  extracted: {
                    documentCategory: 'document',
                    totalCost: 20000,
                  },
                  sourceText:
                    'Retail installment contract with lender payment of $425.50 monthly at 5.9% APR.',
                },
              },
            ],
          },
        ],
      },
    ];

    const insights = computeOwnershipInsights(categories, {
      year: currentYear - 1,
    });

    expect(insights.analyzedDocumentCount).toBe(1);
    expect(insights.financeDocsCount).toBe(1);
    expect(insights.estimatedMonthlyPayment).toBe(425.5);
    expect(insights.estimatedPrincipal).toBe(20000);
    expect(insights.estimatedCurrentValue).toBe(17000);
    expect(insights.estimatedValueRealized).toBe(3000);
    expect(insights.estimatedPaidToDate).toBe(5106);
    expect(insights.upcomingPaymentDates).toHaveLength(6);
  });

  it('uses the actual purchase date, not model year, for loan tenure on a used vehicle', () => {
    // Regression: a used vehicle's loan tenure was computed from its model
    // year, not when it was actually purchased/financed — a 10-year-old
    // model bought (and financed) 3 months ago would previously be treated
    // as having 10 years of payments behind it, overstating
    // estimatedPaidToDate up to the full loan principal (falsely showing
    // the loan as paid off).
    const currentYear = new Date().getFullYear();
    const threeMonthsAgo = new Date();
    threeMonthsAgo.setMonth(threeMonthsAgo.getMonth() - 3);

    const categories = [
      {
        key: 'finance',
        items: [
          {
            files: [
              {
                analysis: {
                  extracted: { documentCategory: 'document', totalCost: 20000 },
                  sourceText: 'Lender payment of $425.50 monthly due on the 1st.',
                },
              },
            ],
          },
        ],
      },
    ];

    const insights = computeOwnershipInsights(categories, {
      year: currentYear - 10,
      purchaseDate: threeMonthsAgo.toISOString().slice(0, 10),
    });

    expect(insights.estimatedPrincipal).toBe(20000);
    // 3 months of $425.50 payments, nowhere near the full principal.
    expect(insights.estimatedPaidToDate).toBeLessThan(2000);
    expect(insights.estimatedPaidToDate).not.toBe(20000);
  });

  it('aggregates maintenance spend and keeps the latest service date', () => {
    const categories = [
      {
        key: 'maintenance',
        items: [
          {
            files: [
              {
                analysis: {
                  extracted: {
                    documentCategory: 'invoice',
                    serviceType: 'Oil change',
                    totalCost: 89.99,
                    serviceDate: '2026-01-15',
                  },
                },
              },
              {
                analysis: {
                  extracted: {
                    documentCategory: 'receipt',
                    serviceType: 'Brake service',
                    totalCost: 410.01,
                    serviceDate: '2026-03-01',
                  },
                },
              },
            ],
          },
        ],
      },
    ];

    const insights = computeOwnershipInsights(categories, { year: 2022 });

    expect(insights.analyzedDocumentCount).toBe(2);
    expect(insights.maintenanceDocsCount).toBe(2);
    expect(insights.maintenanceTotalCost).toBe(500);
    expect(insights.maintenanceAverageCost).toBe(250);
    expect(insights.latestServiceDate).toBe('2026-03-01');
    expect(insights.financeDocsCount).toBe(0);
  });

  it('does not fabricate finance projections without finance evidence', () => {
    const categories = [
      {
        key: 'ownership',
        items: [
          {
            files: [
              {
                analysis: {
                  extracted: {
                    documentCategory: 'document',
                    totalCost: 125,
                  },
                  sourceText: 'Renewal completed for annual registration.',
                },
              },
            ],
          },
        ],
      },
    ];

    const insights = computeOwnershipInsights(categories, { year: 2020 });

    expect(insights.financeDocsCount).toBe(0);
    expect(insights.estimatedMonthlyPayment).toBeUndefined();
    expect(insights.estimatedPrincipal).toBeUndefined();
    expect(insights.estimatedCurrentValue).toBeUndefined();
    expect(insights.upcomingPaymentDates).toEqual([]);
  });

  it('excludes the vehicle purchase price (Bill of Sale) from maintenance spend even though the AI tags it a "receipt"', () => {
    const categories = [
      {
        key: 'ownership',
        items: [
          {
            files: [
              {
                analysis: {
                  extracted: {
                    // Same generic tag a real maintenance receipt gets —
                    // this is the exact ambiguity that caused a $100k+
                    // vehicle purchase price to be counted as "maintenance
                    // spend captured" for a real user's vehicle.
                    documentCategory: 'receipt',
                    totalCost: 108000,
                    serviceDate: '2025-06-01',
                  },
                },
              },
            ],
          },
        ],
      },
      {
        key: 'maintenance',
        items: [
          {
            files: [
              {
                analysis: {
                  extracted: {
                    documentCategory: 'receipt',
                    serviceType: 'Oil change',
                    totalCost: 75,
                    serviceDate: '2026-01-15',
                  },
                },
              },
            ],
          },
        ],
      },
    ];

    const insights = computeOwnershipInsights(categories, { year: 2025 });

    expect(insights.analyzedDocumentCount).toBe(2);
    expect(insights.maintenanceDocsCount).toBe(1);
    expect(insights.maintenanceTotalCost).toBe(75);
    expect(insights.latestServiceDate).toBe('2026-01-15');
  });
});

File: packages/mobile/test/ownership_insights_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:vehicle_vitals_flutter/utils/ownership_insights.dart';

Map<String, dynamic> _file({
  required String documentCategory,
  String? categoryKey,
  String? serviceType,
  double? totalCost,
  String? serviceDate,
  String? sourceText,
  double confidence = 0.8,
}) {
  return {
    'path': 'p/$documentCategory',
    'categoryKey': categoryKey,
    'analysis': {
      'confidence': confidence,
      'sourceText': sourceText,
      'extracted': {
        'documentCategory': documentCategory,
        'serviceType': ?serviceType,
        'totalCost': ?totalCost,
        'serviceDate': ?serviceDate,
      },
    },
  };
}

void main() {
  test('returns empty insights when no files are analyzed', () {
    final insights = computeOwnershipInsights([]);
    expect(insights.hasAnyInsight, isFalse);
    expect(insights.analyzedDocumentCount, 0);
  });

  test('aggregates maintenance docs cost, count, and latest service date', () {
    final files = [
      _file(
        documentCategory: 'receipt',
        categoryKey: 'maintenance',
        serviceType: 'oil change',
        totalCost: 80,
        serviceDate: '2026-01-15',
      ),
      _file(
        documentCategory: 'invoice',
        categoryKey: 'maintenance',
        serviceType: 'brake service',
        totalCost: 320,
        serviceDate: '2026-03-01',
      ),
    ];

    final insights = computeOwnershipInsights(files);

    expect(insights.hasAnyInsight, isTrue);
    expect(insights.analyzedDocumentCount, 2);
    expect(insights.maintenanceDocsCount, 2);
    expect(insights.maintenanceTotalCost, 400);
    expect(insights.maintenanceAverageCost, 200);
    expect(insights.latestServiceDate, '2026-03-01');
  });

  test('detects finance documents and estimates monthly payment', () {
    final files = [
      _file(
        documentCategory: 'document',
        categoryKey: 'finance',
        totalCost: 450,
        sourceText: 'Monthly payment of \$450/mo due on the 1st',
      ),
    ];

    final insights = computeOwnershipInsights(files, vehicleYear: 2022);

    expect(insights.financeDocsCount, 1);
    expect(insights.estimatedMonthlyPayment, 450);
  });

  test('estimates current value and value realized from a large principal', () {
    final currentYear = DateTime.now().year;
    final files = [
      _file(
        documentCategory: 'document',
        categoryKey: 'finance',
        totalCost: 20000,
      ),
    ];

    final insights = computeOwnershipInsights(
      files,
      vehicleYear: currentYear - 2,
    );

    expect(insights.estimatedCurrentValue, isNotNull);
    expect(insights.estimatedValueRealized, isNotNull);
    expect(insights.estimatedCurrentValue! < 20000, isTrue);
  });

  test(
    'excludes the vehicle purchase price (Bill of Sale) from maintenance '
    'spend even though the AI tags it a generic "receipt"',
    () {
      final files = [
        // Same generic tag a real maintenance receipt gets — this is the
        // exact ambiguity that caused a $100k+ vehicle purchase price to be
        // counted as "maintenance spend captured" for a real user's vehicle.
        _file(
          documentCategory: 'receipt',
          categoryKey: 'ownership',
          totalCost: 108000,
          serviceDate: '2025-06-01',
        ),
        _file(
          documentCategory: 'receipt',
          categoryKey: 'maintenance',
          serviceType: 'oil change',
          totalCost: 75,
          serviceDate: '2026-01-15',
        ),
      ];

      final insights = computeOwnershipInsights(files);

      expect(insights.analyzedDocumentCount, 2);
      expect(insights.maintenanceDocsCount, 1);
      expect(insights.maintenanceTotalCost, 75);
      expect(insights.latestServiceDate, '2026-01-15');
    },
  );
}
